### DISCUSSION TITLE
CSI plug-in restarts unexpectedly. How to obtain information about previously managed volumes after the restart

### DISCUSSION THREAD

**Discussion by lijhnihaoa** [ceph/ceph-csi#2888] [Q&A] [answered]
The CSI plug-in restarted unexpectedly, and the volume information created before will be lost. I would like to know how to solve this problem. Or should I look at that part of the code. please let me know, thank you

**Comment by Madhu-1** [comment, marked as answer]
>volume information created before will be lost.

What do you mean by volume information created before lost? can you be more specific with exact problem?

**Reply by lijhnihaoa** [reply]
What I am doing is spdk-csi. After applying for a volume, the controller and node parts of csi will save the id of the volume, which is used to delete the volume later, but if the controller and node of csi restart accidentally, they will be used to identify the deletion of the corresponding volume. information will be lost。
I don't know if what I said can make you understand

**Reply by lijhnihaoa** [reply]
The information used to delete or unmount the volume and its related information will be lost, I want to know how ceph-csi does it

**Reply by Madhu-1** [reply]
>What I am doing is spdk-csi. After applying for a volume, the controller and node parts of csi will save the id of the volume, which is used to delete the volume later, but if the controller and node of csi restart accidentally, they will be used to identify the deletion of the corresponding volume. information will be lost。
I don't know if what I said can make you understand

@lijhnihaoa we don't save any information in the in-memory of the csi driver its returned back to the CO, it's CO's responsibility to send the proper information for all the operations.

**Reply by Madhu-1** [reply]
> The information used to delete or unmount the volume and its related information will be lost, I want to know how ceph-csi does it

We store some information for unmounting on the hostPath. even if the driver is restarted the information will still be available

**Reply by lijhnihaoa** [reply]
How is the uninstall information stored in the hostPath, through the output file? How to ensure the accuracy of the output file? If I accidentally restart before output, this is where I am more worried. Where is the code for this part of the implementation, can you tell me? Thank you so much.

**Reply by Madhu-1** [reply]
metadata will be stored on each individual stagingPath in a file and each volume will have its own metadata files 
https://github.com/ceph/ceph-csi/blob/4235178f7c53ee3e6eba8e37cb175623fc0a8cea/internal/rbd/nodeserver.go#L250-L261
https://github.com/ceph/ceph-csi/blob/4235178f7c53ee3e6eba8e37cb175623fc0a8cea/internal/rbd/nodeserver.go#L333-L337

**Reply by lijhnihaoa** [reply]
I'll take a look at this part of the code first, and then I'll discuss with you if I still have questions. thanks a lot for your answe.

**Reply by lijhnihaoa** [reply]
In the `controller`, how `DeleteVolume` saves the information it needs, I don't understand this place, please tell me.
